### PR TITLE
Only show a notification on map and geostory Save action #134

### PR DESCRIPTION
**Pull request by MuhweziDeo**
Remove the dialog and just show a success (or error) notification when the request has completed.

### CODE REVIEW

File: geonode_mapstore_client/client/js/plugins/save/SaveModal.jsx [hunk 1]
```diff
@@ -39,6 +39,11 @@ function SaveModal({
     const [description, setDescription] =  useState('');
     const [nameValidation, setNameValidation] =  useState(false);
 
+    const checkIsEditMode = () => {
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/plugins/save/SaveModal.jsx:42]: Why do we need to modify this component? Could we directly remove the save modal from the Save plugin?

File: geonode_mapstore_client/client/js/epics/__tests__/gnsave-test.js [hunk 2]
```diff
@@ -44,7 +44,7 @@ describe('gnsave epics', () => {
     it('should create new map with success (gnSaveContent)', (done) => {
         const NUM_ACTIONS = 3;
         const metadata = {
-            title: 'Title',
+            name: 'Title',
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/epics/__tests__/gnsave-test.js:47]: Why do we need to change title to name?

File: geonode_mapstore_client/client/js/epics/gnsave.js [hunk 3]
```diff
@@ -126,23 +128,33 @@ const SaveAPI = {
 };
 
 export const gnSaveContent = (action$, store) =>
-    action$.ofType(SAVE_CONTENT)
+    action$.ofType(SAVE_CONTENT, TOGGLE_CONTROL)
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/epics/gnsave.js:131]: I would prefer if we use a separate epic to manage the toggle control or directly trigger the save content from the menu item of burger menu

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

**Review by allyoucanmap — CHANGES_REQUESTED** [review]
We need to refactor a bit the component:
- Save should not contain the Modal and we could replace it with a loader overlay to make te user aware of the saving status

example
```js
function Save(props) {
    return props.saving ? (<div
        style={{ position: 'absolute', width: '100%', height: '100%', backgroundColor: 'rgba(255, 255, 255, 0.75)' }}
    >
        <Loader size={150}/>
    </div>) : null;
}

const SavePlugin = connect(
    createSelector([
        state => state?.gnsave?.saving
    ], (saving) => ({
        saving
    }))
)(Save);

export default createPlugin('Save', {
    component: SavePlugin,
    containers: {
        BurgerMenu: {
            name: 'save',
            position: 30,
            text: <Message msgId="save"/>,
            icon: <Glyphicon glyph="floppy-open"/>,
            action: saveDirectContent,
            selector: createSelector(
                isLoggedIn,
                state => state?.gnresource?.isNew,
                state => state?.gnresource?.permissions?.canEdit,
                mapInfoSelector,
                (loggedIn, isNew, canEdit, mapInfo) => ({
                    // we should add permList to map pages too
                    // currently the canEdit is located inside the map info
                    style: loggedIn && !isNew && (canEdit || mapInfo?.canEdit) ? {} : { display: 'none' }
                })
            )
        }
    },
    epics: {
        ...gnsaveEpics
    },
    reducers: {
        gnresource,
        gnsave,
        controls
    }
});
```

- restore the save modal component as before because the Save As should work as before
- we should isolate the direct saving content from the current one to perform the missing request. please restore the saveContent epic as before

example

```js
export const gnSaveDirectContent = (action$, store) =>
    action$.ofType(SAVE_DIRECT_CONTENT)
        .switchMap(() => {
            const state = store.getState();
            const mapInfo = mapInfoSelector(state);
            const resourceId = mapInfo.id // injected map id
            || state?.gnresource?.id; // injected geostory id
            return Observable.defer(() => getResourceByPk(resourceId))
                .switchMap((resource) => {
                    const metadata = {
                        name: resource.title,
                        description: resource.abstract,
                        thumbnail: resource.thumbnail_url
                    };
                    return Observable.of(
                        setResource(resource),
                        saveContent(resourceId, metadata, false, true /* showNotification */)
                    );
                })
                .catch((error) => {
                    return Observable.of(errorNotification({}));
                });
        })
        .startWith(savingResource());
```

File: geonode_mapstore_client/client/js/api/geonode/adapter/index.js [hunk 4]
```diff
@@ -43,7 +43,7 @@ export const creatMapStoreMap = (body = {}) => {
 */
 export const updateMapStoreMap = (id, body = {}) => {
     const { endpointAdapter = '/mapstore/rest' } = getConfigProp('geoNodeApi') || {};
-    return axios.patch(parseDevHostname(`${endpointAdapter}/resources/${id}/`),
+    return axios.put(parseDevHostname(`${endpointAdapter}/resources/${id}/`),
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/api/geonode/adapter/index.js:46]: Why do we need to change this to put?

File: geonode_mapstore_client/client/js/epics/gnsave.js [hunk 5]
```diff
@@ -138,14 +142,42 @@ export const gnSaveContent = (action$, store) =>
...
+                        description: resource?.abstract,
+                        thumbnail: resource?.thumbnail_url
+                    };
+                    return Observable.of(
+                        setResource(resource),
+                        saveContent(resourceId, metadata, false, true /* showNotification */)
+                    );
+                })
+                .catch((error) => {
+                    return Observable.of(saveError(error.data || error.message));
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/epics/gnsave.js:178]: We need to show a notification also in this case

File: geonode_mapstore_client/client/js/epics/gnsave.js [hunk 6]
```diff
@@ -138,14 +142,42 @@ export const gnSaveContent = (action$, store) =>
...
-                        })
+                        }),
+                        action.showNotifications && successNotification({title: "saveDialog.saveSuccessTitle", message: "saveDialog.saveSuccessMessage"})
                     );
                 })
                 .catch((error) => {
-                    return Observable.of(saveError(error.data || error.message));
+                    return Observable.of(
+                        saveError(error.data || error.message),
+                        action.showNotifications && errorNotification({title: "errorTitleDefault", message: "errorDefault"})
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/epics/gnsave.js:152]: Are `errorTitleDefault` and `errorDefault` valid message id from MapStore or  they are new?

For the message we could use `map.mapError.errorDefault` https://github.com/GeoNode/geonode-mapstore-client/blob/master/geonode_mapstore_client/client/js/plugins/save/SaveModal.jsx#L100

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

File: geonode_mapstore_client/client/js/api/geonode/adapter/index.js (hunk 4 above)
Comment (MuhweziDeo) [reply geonode_mapstore_client/client/js/api/geonode/adapter/index.js:46]: So Django sets a header for content-type to `text/plain, */*` which causes the request to fail for `patch`. I tried to see how to fix and the only way I could get is to change the http method to `put`.
Screenshot 2021-04-27 at 14 16 03
Screenshot 2021-04-27 at 14 16 10

Comment (MuhweziDeo) [reply geonode_mapstore_client/client/js/api/geonode/adapter/index.js:46]: Okay

Comment (MuhweziDeo) [reply geonode_mapstore_client/client/js/api/geonode/adapter/index.js:46]: it turns out there was a problem in my config. I have restored the patch method

**Pull request merged**